load/app/context.go

```diff
@@ -0,0 +1,211 @@
+	FundAccounts(accounts []common.Address, value *big.Int) error
+	Close()
+}
+
+func NewContext(rpcClient rpc.RpcClient, treasure *Account) (*appContext, error) {
+	// Install the helper contract used by this contract for its operations.
+	// Create a context to interact with the network.
+	res := &appContext{
+		rpcClient: rpcClient,
+		treasure:  treasure,
+	}
+
+	// Install a helper contract on the network to perform operations.
+	helper, receipt, err := DeployContract(res, contract.DeployHelper)
+	if err != nil {
+		return nil, fmt.Errorf("failed to deploy helper contract: %w", err)
+	}
+	if receipt.Status != types.ReceiptStatusSuccessful {
+		return nil, fmt.Errorf("failed to deploy helper contract: transaction reverted")
+	}
+
+	res.helper = helper
+	return res, nil
+}
+
+type appContext struct {
+	rpcClient rpc.RpcClient    // < access to the network
+	treasure  *Account         // < the account paying for management tasks
+	helper    *contract.Helper // < a contract used for on-chain operations
+}
+
+func (c *appContext) Close() {
+	c.rpcClient.Close()
+}
+
+func (c *appContext) GetClient() rpc.RpcClient {
+	return c.rpcClient
+}
+
+func (c *appContext) GetTreasure() *Account {
+	return c.treasure
+}
+
+// GetTransactOptions provides transaction options to be used to send a transaction
+// with the given account. The options include the chain ID, a suggested gas price,
+// the next free nonce of the given account, and a hard-coded gas limit of 1e6.
+// The main purpose of this function is to provide a convenient way to collect all
+// the necessary information required to create a transaction in one place.
+func (c *appContext) GetTransactOptions(account *Account) (*bind.TransactOpts, error) {
+	client := c.rpcClient
+
+	ctxt := context.Background()
+	chainId, err := client.ChainID(ctxt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get chain ID: %w", err)
+	}
+
+	gasPrice, err := client.SuggestGasPrice(ctxt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get gas price suggestion: %w", err)
+	}
+
+	nonce, err := client.NonceAt(ctxt, account.address, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get nonce: %w", err)
+	}
+
+	txOpts, err := bind.NewKeyedTransactorWithChainID(account.privateKey, chainId)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create transaction options: %w", err)
+	}
+	txOpts.GasPrice = new(big.Int).Mul(gasPrice, big.NewInt(2))
+	txOpts.Nonce = big.NewInt(int64(nonce))
+	return txOpts, nil
+}
+
+// GetReceipt waits for the receipt of the given transaction hash to be available.
+// The function times out after 10 seconds.
+func (c *appContext) GetReceipt(txHash common.Hash) (*types.Receipt, error) {
+	client := c.rpcClient
+
+	// Wait for the response with some exponential backoff.
+	const maxDelay = 100 * time.Millisecond
+	begin := time.Now()
+	delay := time.Millisecond
+	for time.Since(begin) < 10*time.Second {
```

Comment:
we do have `retry` utility, which could be used here, exmple: 
https://github.com/Fantom-foundation/Norma/blob/main/driver/docker/docker.go#L201-L205

it is waiting constant dealy, but maybe it is not a problem

Reply:
Yes, but as long as this feature is not factored out I would not use it here. There is no reason why the app package should have a dependency to the docker package.

Reply:
No ... the use case is not fitting ... it should be an exponential retry-mechanism to minimize waiting time in this situation. I actually think the network's `Retry` should be updated as well to use exponential back-off and a maximum waiting time.

Reply:
I agree that exponential back-off is better also in use cases for `Retry`, but I think once we update this method, no one will go back here to change this code, i.e. perhaps using the method here now, is a better trade-off